### DISCUSSION TITLE
How can I update and show proper value in multiple input

### DISCUSSION THREAD

**Discussion by ShekhSaifuddin007** [livewire/livewire#1846] [1. Help] [answered]
### Description
I have a full-page livewire component and I'm doing crud with a single modal, I'm creating variation for a product and variation has multiple values so in that case, I created a one to many relations between Variation and VariationValue. When I create there have no problem but when I want to update there have a problem, my multiple inputs have an index value not showing properly inside inputs, if I doing like $value->value then I see the proper value inside inputs, but I can't update or delete them until if I have no id.

I want that user can update and delete and also add more fields as they wish, but my problem I can't show the proper value inside inputs and can't grab the id to delete from `db`, I'm trying to do with `$editMode like wire:model.defer="{{ $editMode ? $values[$key]->value : 'values.'.$key }}"` but it does not work, it just shows me the value inside` wire:model.defer="value"` when I inspect the element. How can I do this for update and delete also for new values, I searched a lot but no result found, below the image you can see that there have `[object Object]` inside those inputs not showing the proper values.

### When `$editMode = true` it's showing me like this

![form](https://user-images.githubusercontent.com/48278861/96495897-59f51780-126a-11eb-8fd0-2d8491804876.PNG)

### In my modal blade

```html
<div class="col-12">
   <fieldset>
     <label>Variation values *</label>
     @foreach ($moreValues as $key => $value)
        <div class="{{ $loop->last ? '' : 'mb-2' }}" wire:key="{{ $key }}">

          <div class="input-group">
            <input type="text" 
                   wire:model.defer="values.{{ $key }}" 
                   class="form-control"                                       
                   placeholder="Variation value">

            <div class="input-group-append">
             @if ($loop->index === 0)
               <button wire:click.prevent="addRowForValue" 
                       class="btn btn-primary font-size-large" type="button">
                    +
               </button>
             @else
               <button wire:click.prevent="removeRowForValue({{ $key }})"
                       class="btn btn-danger font-size-large" type="button">
                    -
               </button>
             @endif
           </div>
         </div>

        @error("values.{{ $key }}")
           <span class="text-danger font-size-base">{{ $message }}</span>
        @enderror
      </div>
    @endforeach
  </fieldset>
</div>
```

### In my Component

```php
<?php

namespace App\Http\Livewire\Products\Variations;

use App\Models\Variation;
use Livewire\Component;

class Index extends Component
{

    public $moreValues = [0];

    public $editMode = false;

    public $name, $values = [];

    protected $rules = [
        'name' => 'required|string',
        'values.*' => 'required|string',
    ];

    public function addRowForValue()
    {
        $this->moreValues[] += 1;
    }

    public function removeRowForValue($key)
    {
        unset($this->moreValues[$key]);
    }

    public function store()
    {
        $this->validate();

        $variation = Variation::create([
            'name' => $this->name
        ]);

        foreach ($this->values as $key => $value) {
            $variation->values()->create([
                'value' => $this->values[$key]
            ]);
        }

        session()->flash('message', 'Unit create successfully.');

        $this->clearForm();
    }

    /**
     * @param $id
     */
    public function edit($id)
    {
        $this->editMode = true;

        $this->moreValues = [0];

        $this->resetErrorBag();

        $this->clearForm();

        // fill the form here with data
        $variation = Variation::query()->findOrFail($id);

        $this->name = $variation->name;

        $this->values = [];

        foreach ($variation->values as $key => $value) {

            // incrementing input by value
            $this->moreValues[$key] = $value;

            // showing value their own field
            $this->values[$key] = $value;
        }

        $this->emit('openModal');
    }

    protected function clearForm()
    {
        $this->name = '';
        $this->values = '';
    }

    public function render()
    {
        $variations = Variation::with('values')->latest('id')->get();

        return view('livewire.products.variations.index', compact('variations'))
            ->extends('layouts.app');
    }
}
```

**Context**
- Livewire version: 2.x
- Laravel version: 8.x
- Browser: all

**Comment by MichaelBrauner** [comment]
I guess the problem is that you bind an object with multiple Variation values to an input field. 
And Livewire don't know, how to put this into the field. 

You can access the livewire model like this: 

```
@props([
'modelName'=> $attributes->wire('model')->value()
])
```

Afterwards you can iterate over the values with plain blade:

```
@foreach($this->$modelName as $item) 
   {{--  do some stuff --}}
@endforeach
```

Hope this helps.

**Comment by ShekhSaifuddin007** [comment]
@Slowwie actually my modal is in the same component, I just want value in the input fields. When it's editMode. It is a same modal for edit and update. It is just check the `$editMode` variable true or false....

**Comment by MichaelBrauner** [comment]
Can you dumb the object, to see, whats inside it? I kind of tap a bit in the dark.

**Comment by ShekhSaifuddin007** [comment]
@Slowwie thank you for your response brother I changed my code a little bit and now it's fine I remove `public $moreValues = [0];` this property to increment my fields, directly use `public $values = [0];` and add db column name 'value' now it showing properly what I need here is a demo

![form2](https://user-images.githubusercontent.com/48278861/96615086-a8fe8380-1322-11eb-8cf3-c5c6f4796b42.PNG)


### In my blade look like

```html
<div class="col-12">
   <fieldset>
     <label>Variation values *</label>
     @foreach ($values as $key => $value)
        <div class="{{ $loop->last ? '' : 'mb-2' }}" wire:key="{{ $key }}">

          <div class="input-group">
            <input type="text" 
                   wire:model.defer="values.{{ $key }}.value" 
                   class="form-control"                                       
                   placeholder="Variation value">

            <div class="input-group-append">
             @if ($loop->index === 0)
               <button wire:click.prevent="addRowForValue" 
                       class="btn btn-primary font-size-large" type="button">
                    +
               </button>
             @else
               <button wire:click.prevent="removeRowForValue({{ $key }})"
                       class="btn btn-danger font-size-large" type="button">
                    -
               </button>
             @endif
           </div>
         </div>

        @error("values.{{ $key }}.value")
           <span class="text-danger font-size-base">{{ $message }}</span>
        @enderror
      </div>
    @endforeach
  </fieldset>
</div>
```

**Comment by MichaelBrauner** [comment]
Nice! You fixed it! ;)

**Comment by ShekhSaifuddin007** [comment]
@Slowwie this issue is fixed but another comes now I have to give functionality that the user can update the fields and add more also and if they wish to delete some they can, but I can't do this, you can see in my code that inside update function I'm trying to create a new record by checking via `$id` but not going in `if` condition it just jumps to `else` and update all existing record not creating new. Here is my code

```php
public function update()
{
        $this->validate([
            'name' => "required|string|unique:variations,name,{$this->variation_id}",
            'values.*.value' => 'required|string',
        ]);

        $variation = Variation::query()->findOrFail($this->variation_id);
        $variation->update(['name' => $this->name]);


        foreach ($this->values as $key => $value) {
            $id = $this->values[$key]['id'];
            $variationValue = VariationValue::query()->findOrFail($id);

            if (! $id) {
                $variation->values()->create([
                    'value' => $this->values[$key]['value']
                ]);
            } else {
                $variationValue->update([
                    'value' => $this->values[$key]['value']
                ]);
           }
     }


     session()->flash('message', 'Variation updated successfully.');

     $this->clearForm();

     $this->emit('closeModal');
}
```

How can I do this I don't understand actually. what is wrong in my code I can't figure out.

**Comment by danharrin** [comment]
Hi, please post a link to a Laravel Playground so that we can more easily reproduce your issue. Many thanks!

**Comment by MichaelBrauner** [comment]
When it jumps to 

```
} else {
     $variationValue->update([
          'value' => $this->values[$key]['value']
     ]);
}
```

I would `dump($id)` before the if statement  - so I can figure out why it jumps to the else.

**Comment by ShekhSaifuddin007** [comment]
@Slowwie yes I `dump($id)`, when I dump it shows me id but nothing happen..😢

**Comment by ShekhSaifuddin007** [comment]
@danharrin I'm tried to create a playground, but there have no relationship opportunities in the livewire preset where I can go with relation and create data. 😒

[here](https://laravelplayground.com/#/snippets/80e6df2b-e2b8-40b9-a619-0c90b73de53a)

**Comment by danharrin** [comment]
@ShekhSaifuddin007 take a look at the Eloquent example there, you should theoretically be able to replicate the model and migration files in the Livewire example.

**Comment by ShekhSaifuddin007** [comment]
@danharrin can you see this now I create a playground, it will be very helpful for me, I can update every record but when I want to create some new record I this not go my `if` part, what I doing wrong there I can't understand actually. I want that users can create new records and if they wish to delete some they can but I'm unable to do this.😒

[here](https://laravelplayground.com/#/snippets/0c396390-4f7a-4baf-b36a-dbcb968d4f14)

**Comment by ShekhSaifuddin007** [comment]
@danharrin and @Slowwie brothers look into that, I really need your help, This is my running project for a client with livewire. please

**Comment by danharrin** [comment, marked as answer]
```php
foreach ($this->values as $key => $value) {
    $id = $this->values[$key]['id'] ?? false;

    if ($id) {
        VariationValue::findOrFail($id)->update([
            'value' => $this->values[$key]['value']
        ]);
    } else {
        $variation->values()->create([
            'value' => $this->values[$key]['value']
        ]);
    }
}
```

@ShekhSaifuddin007

**Reply by ShekhSaifuddin007** [reply]
@danharrin yes this work....✌✌, thank you brother..🧡❤💛🧡❤💛, you save my life

**Reply by danharrin** [reply]
You’re welcome ❤️
